### ISSUE TITLE
Support lua funcref in `keywordprg`

### ISSUE THREAD

**Issue by clason** [neovim/neovim#18997]
### Feature already in Vim?

N/A

### Feature description

Differently from other `fooprg` options, `keywordprg` (used for `K` command) supports an ex command:
```
	When the first character is ":", the command is invoked as a Vim
	command prefixed with [count].
```

It would be useful to be able to specify (via the Neovim API from Lua, or the Lua API, only) a Lua funcref.

Use case: setting `keywordprg` to `vim.lsp.buf.hover` to avoid having to remap `K` conditionally.

**Comment by brianhuster** [comment]
> Use case: setting keywordprg to vim.lsp.buf.hover (or an appropriate wrapper) to avoid having to remap K conditionally.

I still want to use both ftplugin's keywordprg and LSP hover, so I map this `nnoremap <C-k> K`. And as this comment https://github.com/neovim/neovim/pull/24331#issuecomment-1636107198 says, I can also use `v_K`

But setting keywordprg to use LSP will make those tricks no longer work

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/41065 implements this for "func" and "expr" options. 

"prg" options are less obvious but worth doing. Unlike "expr" options (which often are evaluated on a hot path, e.g. `foldexpr`), we can be less performance-sensitive with the approach to "prg" options: the legacy string form can be implemented as a Lua closure. So we can just always store Lua closures for "prg", which is helpful because they have weird, bespoke semantics which is difficult to express at the codegen level.

**Comment by 616b2f** [comment]
does this mean that we can assign a lua function to `makeprg` after that too?

**Comment by justinmk** [comment]
> does this mean that we can assign a lua function to `makeprg` after that too?

yes, though the details need to be figured out. my assumption is that each function will need option-specific parameters.

**Comment by clason** [comment]
Note that most "prg" options take direct shell commands; it's not always clear how a Nvim function would/should work. There's no technical barrier, but the exact semantics would have to be discussed on a case-by-case basis.
